<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query = f'''
select 
   count(*)
from 
   public.vacancies
        '''

In [5]:
df  = pd.read_sql_query(query, connection)

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [6]:
# результат запроса
49197

49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [7]:
# текст запроса
query_1 = f'''
select distinct(id)
from public.employers
'''
df_em_co = pd.read_sql_query(query_1, connection)
df_em_co.shape[0]

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


23501

In [8]:
# результат запроса
23501

23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [9]:
# текст запроса
query_2 = f'''
select distinct *
from 
    public.areas
'''
df_reg_co = pd.read_sql_query(query_2, connection)
df_reg_co.shape[0]

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


1362

In [10]:
# результат запроса
1362

1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [11]:
# текст запроса
query_3 = f'''
select *
from 
    public.industries
'''
df_ind_co = pd.read_sql_query(query_3, connection)
df_ind_co.shape[0]

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


294

In [12]:
# результат запроса
294

294

***

In [13]:
# выводы по предварительному анализу данных

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [14]:
# текст запроса
query_4_1 = f'''
select
    v.id,
    a.name
from 
    public.vacancies v
join public.areas a on v.area_id = a.id
'''

df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1['name'].value_counts(normalize=True)

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Москва                            0.108401
Санкт-Петербург                   0.057951
Минск                             0.042929
Новосибирск                       0.040775
Алматы                            0.038458
                                    ...   
Тура (Красноярский край)          0.000020
Новоульяновск                     0.000020
Абхазия                           0.000020
Заполярный (Ямало-Ненецкий АО)    0.000020
Изобильный                        0.000020
Name: name, Length: 769, dtype: float64

# результат запроса
Москва                            0.108401
Санкт-Петербург                   0.057951
Минск                             0.042929
Новосибирск                       0.040775
Алматы                            0.038458

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_4_2 = f'''
select
    id,
    salary_from,
    salary_to
from
    public.vacancies
where
    salary_from is not NULL
    or salary_to is not NULL
group by 1
'''


df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2.info

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


<bound method DataFrame.info of              id  salary_from  salary_to
0      55215815      40000.0    60000.0
1      55101356      42716.0    54921.0
2      55563348          NaN   200000.0
3      55412764          NaN   160000.0
4      55301926     150000.0        NaN
...         ...          ...        ...
24068  49640363      80000.0    80000.0
24069  55151414      38500.0    46500.0
24070  54935722      33000.0    60000.0
24071  55035784      80000.0        NaN
24072  55326850      58000.0        NaN

[24073 rows x 3 columns]>

In [16]:
# результат запроса
24073

24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
round(df_4_2['salary_from'].mean())

71065

In [18]:
# результат запроса
round(df_4_2['salary_to'].mean())

110537

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# текст запроса
71065

71065

In [20]:
# результат запроса
110537

110537

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# текст запроса
query_4_5 = f'''
select
    schedule,
    employment
from
    public.vacancies
'''

df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5.value_counts()

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule          employment         
Полный день       Полная занятость       35367
Удаленная работа  Полная занятость        7802
Гибкий график     Полная занятость        1593
Удаленная работа  Частичная занятость     1312
Сменный график    Полная занятость         940
Полный день       Стажировка               569
Вахтовый метод    Полная занятость         367
Полный день       Частичная занятость      347
Гибкий график     Частичная занятость      312
Полный день       Проектная работа         141
Удаленная работа  Проектная работа         133
Гибкий график     Стажировка               116
Сменный график    Частичная занятость      101
Удаленная работа  Стажировка                64
Гибкий график     Проектная работа          18
Сменный график    Стажировка                12
Вахтовый метод    Проектная работа           2
Сменный график    Проектная работа           1
dtype: int64

In [22]:
# результат запроса
query_4_6 = f'''
select
    experience
from
    public.vacancies
'''

df_4_6 = pd.read_sql_query(query_4_6, connection)
df_4_6.value_counts()

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience        
От 1 года до 3 лет    26152
От 3 до 6 лет         14511
Нет опыта              7197
Более 6 лет            1337
dtype: int64

***

# выводы по детальному анализу вакансий
experience        
От 1 года до 3 лет    26152
От 3 до 6 лет         14511
Нет опыта              7197
Более 6 лет            1337
dtype: int64

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [23]:
# текст запроса
query_5 = f'''
select
    e.name
from
    public.vacancies v
join
    public.employers e on e.id = v.employer_id
'''

df_5 = pd.read_sql_query(query_5, connection)
df_5.value_counts()

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name                     
Яндекс                       1933
Ростелеком                    491
Тинькофф                      444
СБЕР                          428
Газпром нефть                 331
                             ... 
Демеш Анна Константиновна       1
Демиург                         1
Демиус                          1
ДенКаРС                         1
“Объединенные сети”             1
Length: 14766, dtype: int64

In [24]:
# результат запроса

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [25]:
# текст запроса
query_5_1 = f'''
select
    a.name,
    count(v.employer_id) as employ_total,
    count(v.id) as vac_total
from
    public.areas a
join
    public.vacancies v on a.id = v.area_id
group by 1
order by 3
'''

df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1.value_counts()


c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name                                 employ_total  vac_total
Абакан                               37            37           1
Малоярославец                        4             4            1
Петрозаводск                         97            97           1
Петропавловка (Воронежская область)  1             1            1
Петропавловка (Республика Бурятия)   1             1            1
                                                               ..
Каменск-Шахтинский                   6             6            1
Камешково                            2             2            1
Камышин                              11            11           1
Камышлов                             1             1            1
Яя                                   1             1            1
Length: 769, dtype: int64

In [26]:
# результат запроса

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [27]:
# текст запроса
query_5_3 = f'''
select
    e.name,
    count(distinct(v.area_id))
from
    public.employers e
left join
    public.vacancies v on e.id = v.employer_id
group by 1
order by 2 desc

'''

df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3.value_counts()

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name                                                                             count
 DANTONE HOME                                                                    1        1
ОКНА PLAST+                                                                      1        1
ОКЕЙ – Федеральная розничная сеть                                                15       1
ОКБ-ПЛАНЕТА                                                                      1        1
ОКБ им.А.С.Яковлева                                                              1        1
                                                                                         ..
ВГТРК (Всероссийская государственная телевизионная и радиовещательная компания)  1        1
ВВС                                                                              0        1
ВВЕРХ ГРУПП                                                                      1        1
ВБЦ                                                                              1   

In [28]:
# результат запроса
181

181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [29]:
# текст запроса
query_5_4 = f'''
select
    count(*)
from public.employers e
left join 
    public.employers_industries i on e.id = i.employer_id
where i.industry_id is null

'''

df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4.value_counts()

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
8419     1
dtype: int64

In [30]:
# результат запроса

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [31]:
# текст запроса
query_5_5 = f'''
select
    e.id,
    e.name,
    count(*)
from public.employers e
left join 
    public.employers_industries i on e.id = i.employer_id
group by 1, 2
having count(*)=4
order by 2
offset 2 limit 1

'''

df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5.value_counts()

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


id     name  count
64174  2ГИС  4        1
dtype: int64

# результат запроса
2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''
select
    count(*)
from public.employers_industries i
left join
    public.industries id on id.id = i.industry_id
where id.name = 'Разработка программного обеспечения'
''' 
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6.value_counts()

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
3553     1
dtype: int64

In [33]:
# результат запроса
3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
lsit_city = page.find_all('table')
lst_city = lsit_city[1]
list_city = []
for i in lst_city.find_all('td'):
    name = i.text
    list_city.append(name)
list_city1 = []
i = 1
while i < len(list_city):
    list_city1.append(list_city[i])
    i+=5


In [35]:
# текст запроса
query_5_7 = f''' 
(select
    a.name as "Город",
    count(v.id)
from public.vacancies v
join
    public.areas a on v.area_id = a.id
join
    public.employers e on v.employer_id = e.id
where e.name = 'Яндекс'
    and a.name in ('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград')
group by 1)

union

(select
    'Total',
    count(v.id)
from public.vacancies v
join
    public.areas a on v.area_id = a.id
join
    public.employers e on v.employer_id = e.id
where e.name = 'Яндекс'
    and a.name in ('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград'))
'''

df_5_7 = pd.read_sql_query(query_5_7, connection)
print(df_5_7)

              Город  count
0             Total    485
1         Волгоград     24
2           Воронеж     32
3      Екатеринбург     39
4            Казань     25
5         Краснодар     30
6        Красноярск     23
7            Москва     54
8   Нижний Новгород     36
9       Новосибирск     35
10             Омск     21
11            Пермь     25
12   Ростов-на-Дону     25
13           Самара     26
14  Санкт-Петербург     42
15              Уфа     26
16        Челябинск     22


c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


# результат запроса
17 строк
Total - 485

***

In [36]:
# выводы по анализу работодателей

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
# текст запроса
query_6_1 = f'''
select
    count(distinct(v.id))
from
    public.vacancies v 
where v.name like '%данн%'
    or v.name like '%data%'
    or v.name like '%Data%'
    or v.name like '%Данн%'
    or v.name like '%ДАНН%'
    or v.name like '%DATA%'
''' 
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1.info

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


<bound method DataFrame.info of    count
0   1771>

In [38]:
# результат запроса
1771

1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
# текст запроса
query_6_2 = f'''
select
    count(v.id)
from
    public.vacancies v 
where (lower(v.name) like '%data scientist%'
    or lower(v.name) like '%data science%'
    or lower(v.name) like '%исследователь данных%'
    or (v.name like '%ML%'
    and v.name not like '%HTML%')
    or lower(v.name) like '%machine learning%'
    or lower(v.name) like '%машинн%обучен%%')
    and (lower(v.name) like '%junior%'
    or v.experience like '%Нет опыта%'
    or v.employment like '%Стажировка%')
''' 
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2.info

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


<bound method DataFrame.info of    count
0     51>

# результат запроса
51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [40]:
# текст запроса
query_6_3 = f'''
select
    count(v.id)
from
    public.vacancies v 
where (lower(v.name) like '%data scientist%'
    or lower(v.name) like '%data science%'
    or lower(v.name) like '%исследователь данных%'
    or (v.name like '%ML%'
    and v.name not like '%HTML%')
    or lower(v.name) like '%machine learning%'
    or lower(v.name) like '%машинн%обучен%%')
    and (v.key_skills like '%SQL%'
    or lower(v.key_skills) like '%postgres%')
''' 
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3.info

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


<bound method DataFrame.info of    count
0    201>

# результат запроса
201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [41]:
# текст запроса
query_6_4 = f'''
select
    count(v.id)
from
    public.vacancies v 
where (lower(v.name) like '%data scientist%'
    or lower(v.name) like '%data science%'
    or lower(v.name) like '%исследователь данных%'
    or (v.name like '%ML%'
    and v.name not like '%HTML%')
    or lower(v.name) like '%machine learning%'
    or lower(v.name) like '%машинн%обучен%%')
    and (lower(v.key_skills) like '%python%')
''' 
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4.info

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


<bound method DataFrame.info of    count
0    351>

# результат запроса
351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [42]:
# текст запроса
query_6_5 = f'''
select
    avg(length(v.key_skills) - length(replace(v.key_skills, CHR(9), ''))) avg
from
    public.vacancies v
where (lower(v.name) like '%data scientist%'
    or lower(v.name) like '%data science%'
    or lower(v.name) like '%исследователь данных%'
    or (v.name like '%ML%'
    and v.name not like '%HTML%')
    or lower(v.name) like '%machine learning%'
    or lower(v.name) like '%машинн%обучен%%')
    and (length(v.key_skills) - length(replace(v.key_skills, CHR(9), '')) !=0)

''' 
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5.info

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


<bound method DataFrame.info of         avg
0  5.456674>

In [43]:
# результат запроса
5.46

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [44]:
# текст запроса
query_6_6 = f'''
select
    distinct(v.experience),
    avg(coalesce(v.salary_from, v.salary_to) + coalesce(v.salary_to, v.salary_from)) / 2
from
    public.vacancies v
where (lower(v.name) like '%data scientist%'
    or lower(v.name) like '%data science%'
    or lower(v.name) like '%исследователь данных%'
    or (v.name like '%ML%'
    and v.name not like '%HTML%')
    or lower(v.name) like '%machine learning%'
    or lower(v.name) like '%машинн%обучен%%')
    and (v.salary_from is not null
    or v.salary_to is not null)
    and (v.experience like '%От 3 до 6 лет%')
group by 1 
''' 
df_6_6 = pd.read_sql_query(query_6_6, connection)
connection.close()
df_6_6.info

c:\Users\R\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


<bound method DataFrame.info of       experience       ?column?
0  От 3 до 6 лет  243114.666667>

# результат запроса
243115

***

In [45]:
# выводы по предметному анализу

# Общий вывод по проекту

In [46]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования